File: oracle-function/switchboard_function.js
#!/usr/bin/env node
/**
 * Switchboard Function for DePIN Activity Verification (Node.js version)
 * This function runs in a Trusted Execution Environment (TEE) to verify:
 * 1. GPS coordinates are valid and realistic
 * 2. IP address is not from a known VPN/proxy service
 * 3. Signal strength is within reasonable bounds for the location
 */

const https = require('https');

class ActivityVerifier {
    constructor() {
        this.vpnProviders = [
            'nordvpn', 'expressvpn', 'surfshark', 'cyberghost', 'protonvpn',
            'mullvad', 'windscribe', 'tunnelbear', 'hotspotshield', 'purevpn'
        ];
    }

    /**
     * Verify GPS coordinates are valid and not obviously fake
     */
    verifyGpsCoordinates(lat, lng) {
        // Basic range validation
        if (lat < -90 || lat > 90 || lng < -180 || lng > 180) {
            return { valid: false, reason: 'GPS coordinates out of valid range' };
        }

        // Check for common fake coordinates (0,0), (null island)
        if (Math.abs(lat) < 0.001 && Math.abs(lng) < 0.001) {
            return { valid: false, reason: 'Suspicious coordinates near null island' };
        }

        // Check for coordinates in the middle of oceans (simplified)
        const oceanAreas = [
            // Pacific Ocean center
            { latRange: [-10, 10], lngRange: [-170, -120] },
            // Atlantic Ocean center  
            { latRange: [-10, 10], lngRange: [-40, -10] },
            // Indian Ocean center
            { latRange: [-20, 0], lngRange: [60, 100] }
        ];

        for (const area of oceanAreas) {
            if (lat >= area.latRange[0] && lat <= area.latRange[1] &&
                lng >= area.lngRange[0] && lng <= area.lngRange[1]) {
                return { valid: false, reason: 'Coordinates appear to be in middle of ocean' };
            }
        }

        return { valid: true, reason: 'GPS coordinates valid' };
    }

    /**
     * Make HTTP request with promise
     */
    makeRequest(url) {
        return new Promise((resolve, reject) => {
            const request = https.get(url, { timeout: 5000 }, (response) => {
                let data = '';
                response.on('data', chunk => data += chunk);
                response.on('end', () => {
                    try {
                        resolve(JSON.parse(data));
                    } catch (e) {
                        reject(new Error('Invalid JSON response'));
                    }
                });
            });

            request.on('error', reject);
            request.on('timeout', () => {
                request.destroy();
                reject(new Error('Request timeout'));
            });
        });
    }

    /**
     * Check if IP address is from VPN/proxy/datacenter
     */
    async checkIpReputation(ipAddress) {
        try {
            const services = [
                `https://ipapi.co/${ipAddress}/json/`,
                `https://ipinfo.io/${ipAddress}/json`
            ];

            for (const serviceUrl of services) {
                try {
                    const data = await this.makeRequest(serviceUrl);
                    
                    // Check for VPN/proxy indicators
                    const org = (data.org || '').toLowerCase();
                    const company = (data.company?.name || '').toLowerCase();
                    
                    // Check against known VPN providers
                    for (const provider of this.vpnProviders) {
                        if (org.includes(provider) || company.includes(provider)) {
                            return { valid: false, reason: `IP associated with VPN provider: ${provider}` };
                        }
                    }
                    
                    // Check for datacenter/hosting providers
                    const datacenterIndicators = ['datacenter', 'hosting', 'cloud', 'server', 'vps', 'aws', 'google', 'azure'];
                    for (const indicator of datacenterIndicators) {
                        if (org.includes(indicator) || company.includes(indicator)) {
                            return { valid: false, reason: `IP from datacenter/hosting provider: ${indicator}` };
                        }
                    }
                    
                    return { valid: true, reason: 'IP reputation check passed' };
                    
                } catch (e) {
                    continue;
                }
            }
            
            return { valid: false, reason: 'Unable to verify IP reputation' };
            
        } catch (e) {
            return { valid: false, reason: `IP reputation check failed: ${e.message}` };
        }
    }

    /**
     * Verify WiFi signal strength is plausible
     */
    verifySignalStrength(signalStrength, lat, lng) {
        // Basic range check (WiFi typically -100 to 0 dBm)
        if (signalStrength < -100 || signalStrength > 0) {
            return { valid: false, reason: `Signal strength ${signalStrength} dBm out of valid range` };
        }

        // Very weak signals might indicate spoofing
        if (signalStrength < -90) {
            return { valid: false, reason: `Signal strength ${signalStrength} dBm too weak, possible spoofing` };
        }

        // Perfect signals are suspicious
        if (signalStrength > -10) {
            return { valid: false, reason: `Signal strength ${signalStrength} dBm suspiciously strong` };
        }

        return { valid: true, reason: 'Signal strength validation passed' };
    }

    /**
     * Calculate distance between two GPS coordinates using Haversine formula
     */
    haversineDistance(lat1, lon1, lat2, lon2) {
        const R = 6371; // Earth's radius in kilometers
        
        const lat1Rad = lat1 * Math.PI / 180;
        const lon1Rad = lon1 * Math.PI / 180;
        const lat2Rad = lat2 * Math.PI / 180;
        const lon2Rad = lon2 * Math.PI / 180;
        
        const dlat = lat2Rad - lat1Rad;
        const dlon = lon2Rad - lon1Rad;
        
        const a = Math.sin(dlat/2) * Math.sin(dlat/2) + 
                  Math.cos(lat1Rad) * Math.cos(lat2Rad) * 
                  Math.sin(dlon/2) * Math.sin(dlon/2);
        const c = 2 * Math.asin(Math.sqrt(a));
        
        return R * c;
    }

    /**
     * Analyze if movement between locations is physically possible
     */
    analyzeMovementPattern(currentLat, currentLng, previousLat = null, previousLng = null, timeDiff = 3600) {
        if (previousLat === null || previousLng === null) {
            return { valid: true, reason: 'No previous location to compare' };
        }

        const distanceKm = this.haversineDistance(previousLat, previousLng, currentLat, currentLng);
        const timeHours = timeDiff / 3600;
        const requiredSpeed = timeHours > 0 ? distanceKm / timeHours : Infinity;
        
        // Maximum reasonable speed (accounting for flights)
        const maxSpeed = 900; // km/h (commercial aircraft speed)
        
        if (requiredSpeed > maxSpeed) {
            return { 
                valid: false, 
                reason: `Impossible movement: ${distanceKm.toFixed(1)}km in ${timeHours.toFixed(1)}h requires ${requiredSpeed.toFixed(1)}km/h` 
            };
        }
        
        return { 
            valid: true, 
            reason: `Movement pattern valid: ${distanceKm.toFixed(1)}km in ${timeHours.toFixed(1)}h` 
        };
    }

    /**
     * Main verification function that combines all checks
     */
    async verifyActivity(activityData) {
        const lat = activityData.gps_lat || 0;
        const lng = activityData.gps_long || 0;
        const signalStrength = activityData.signal_strength || -100;
        const ipAddress = activityData.ip_address || '';
        const previousLat = activityData.previous_lat;
        const previousLng = activityData.previous_lng;
        const timeDiff = activityData.time_diff || 3600;

        const verificationResult = {
            verified: false,
            score: 0,
            max_score: 100,
            checks: {},
            timestamp: Math.floor(Date.now() / 1000)
        };

        // GPS Coordinate Verification (30 points)
        const gpsResult = this.verifyGpsCoordinates(lat, lng);
        verificationResult.checks.gps_validation = {
            passed: gpsResult.valid,
            reason: gpsResult.reason,
            points: gpsResult.valid ? 30 : 0
        };
        if (gpsResult.valid) {
            verificationResult.score += 30;
        }

        // IP Reputation Check (40 points)
        if (ipAddress) {
            const ipResult = await this.checkIpReputation(ipAddress);
            verificationResult.checks.ip_reputation = {
                passed: ipResult.valid,
                reason: ipResult.reason,
                points: ipResult.valid ? 40 : 0
            };
            if (ipResult.valid) {
                verificationResult.score += 40;
            }
        } else {
            verificationResult.checks.ip_reputation = {
                passed: false,
                reason: 'No IP address provided',
                points: 0
            };
        }

        // Signal Strength Verification (20 points)
        const signalResult = this.verifySignalStrength(signalStrength, lat, lng);
        verificationResult.checks.signal_strength = {
            passed: signalResult.valid,
            reason: signalResult.reason,
            points: signalResult.valid ? 20 : 0
        };
        if (signalResult.valid) {
            verificationResult.score += 20;
        }

        // Movement Pattern Analysis (10 points)
        const movementResult = this.analyzeMovementPattern(lat, lng, previousLat, previousLng, timeDiff);
        verificationResult.checks.movement_pattern = {
            passed: movementResult.valid,
            reason: movementResult.reason,
            points: movementResult.valid ? 10 : 0
        };
        if (movementResult.valid) {
            verificationResult.score += 10;
        }

        // Overall verification (require 70% score)
        verificationResult.verified = verificationResult.score >= 70;

        return verificationResult;
    }
}

/**
 * Main function for Switchboard execution
 */
async function main() {
    try {
        let inputData;
        
        // Parse input from command line argument or stdin
        if (process.argv.length > 2) {
            inputData = JSON.parse(process.argv[2]);
        } else {
            // Read from stdin
            const input = await new Promise((resolve) => {
                let data = '';
                process.stdin.on('data', chunk => data += chunk);
                process.stdin.on('end', () => resolve(data));
            });
            inputData = JSON.parse(input);
        }

        // Initialize verifier
        const verifier = new ActivityVerifier();
        
        // Perform verification
        const result = await verifier.verifyActivity(inputData);
        
        // Output result for Switchboard
        const output = {
            success: result.verified,
            data: result
        };
        
        console.log(JSON.stringify(output));
        
        // Return appropriate exit code
        process.exit(result.verified ? 0 : 1);
        
    } catch (error) {
        const errorOutput = {
            success: false,
            error: error.message,
            data: null
        };
        console.log(JSON.stringify(errorOutput));
        process.exit(1);
    }
}

// Run main function if this file is executed directly
if (require.main === module) {
    main();
}

module.exports = { ActivityVerifier };

File: oracle-function/switchboard_function.py
#!/usr/bin/env python3
"""
Switchboard Function for DePIN Activity Verification
This function runs in a Trusted Execution Environment (TEE) to verify:
1. GPS coordinates are valid and realistic
2. IP address is not from a known VPN/proxy service
3. Signal strength is within reasonable bounds for the location
"""

import json
import requests
import sys
import os
from typing import Dict, Any, Tuple
import math

# Switchboard imports
try:
    from switchboard_py import SwitchboardProgram, FunctionRequestAccount
except ImportError:
    print("Warning: Switchboard SDK not available, running in standalone mode")

# Configuration
VPN_DETECTION_API_KEY = os.getenv("VPN_DETECTION_API_KEY", "")
GEOLOCATION_API_KEY = os.getenv("GEOLOCATION_API_KEY", "")

class ActivityVerifier:
    def __init__(self):
        self.vpn_providers = [
            "nordvpn", "expressvpn", "surfshark", "cyberghost", "protonvpn",
            "mullvad", "windscribe", "tunnelbear", "hotspotshield", "purevpn"
        ]
        
    def verify_gps_coordinates(self, lat: float, lng: float) -> Tuple[bool, str]:
        """Verify GPS coordinates are valid and not obviously fake"""
        
        # Basic range validation
        if not (-90 <= lat <= 90) or not (-180 <= lng <= 180):
            return False, "GPS coordinates out of valid range"
        
        # Check for common fake coordinates (0,0), (null island)
        if abs(lat) < 0.001 and abs(lng) < 0.001:
            return False, "Suspicious coordinates near null island"
        
        # Check for coordinates in the middle of oceans (simplified)
        ocean_areas = [
            # Pacific Ocean center
            {"lat_range": (-10, 10), "lng_range": (-170, -120)},
            # Atlantic Ocean center  
            {"lat_range": (-10, 10), "lng_range": (-40, -10)},
            # Indian Ocean center
            {"lat_range": (-20, 0), "lng_range": (60, 100)}
        ]
        
        for area in ocean_areas:
            if (area["lat_range"][0] <= lat <= area["lat_range"][1] and 
                area["lng_range"][0] <= lng <= area["lng_range"][1]):
                return False, "Coordinates appear to be in middle of ocean"
        
        return True, "GPS coordinates valid"
    
    def check_ip_reputation(self, ip_address: str) -> Tuple[bool, str]:
        """Check if IP address is from VPN/proxy/datacenter"""
        
        try:
            # Use multiple IP reputation services
            services = [
                f"https://ipapi.co/{ip_address}/json/",
                f"https://ipinfo.io/{ip_address}/json"
            ]
            
            for service_url in services:
                try:
                    response = requests.get(service_url, timeout=5)
                    if response.status_code == 200:
                        data = response.json()
                        
                        # Check for VPN/proxy indicators
                        org = data.get("org", "").lower()
                        company = data.get("company", {}).get("name", "").lower() if isinstance(data.get("company"), dict) else ""
                        
                        # Check against known VPN providers
                        for provider in self.vpn_providers:
                            if provider in org or provider in company:
                                return False, f"IP associated with VPN provider: {provider}"
                        
                        # Check for datacenter/hosting providers
                        datacenter_indicators = ["datacenter", "hosting", "cloud", "server", "vps", "aws", "google", "azure"]
                        for indicator in datacenter_indicators:
                            if indicator in org or indicator in company:
                                return False, f"IP from datacenter/hosting provider: {indicator}"
                        
                        return True, "IP reputation check passed"
                        
                except Exception as e:
                    continue
            
            return False, "Unable to verify IP reputation"
            
        except Exception as e:
            return False, f"IP reputation check failed: {str(e)}"
    
    def verify_signal_strength(self, signal_strength: int, lat: float, lng: float) -> Tuple[bool, str]:
        """Verify WiFi signal strength is plausible for the location"""
        
        # Basic range check (WiFi typically -100 to 0 dBm)
        if not (-100 <= signal_strength <= 0):
            return False, f"Signal strength {signal_strength} dBm out of valid range"
        
        # Very weak signals might indicate spoofing
        if signal_strength < -90:
            return False, f"Signal strength {signal_strength} dBm too weak, possible spoofing"
        
        # Perfect signals are suspicious
        if signal_strength > -10:
            return False, f"Signal strength {signal_strength} dBm suspiciously strong"
        
        return True, "Signal strength validation passed"
    
    def analyze_movement_pattern(self, current_lat: float, current_lng: float, 
                                previous_lat: float = None, previous_lng: float = None,
                                time_diff: int = 3600) -> Tuple[bool, str]:
        """Analyze if movement between locations is physically possible"""
        
        if previous_lat is None or previous_lng is None:
            return True, "No previous location to compare"
        
        # Calculate distance using Haversine formula
        def haversine_distance(lat1, lon1, lat2, lon2):
            R = 6371  # Earth's radius in kilometers
            
            lat1_rad = math.radians(lat1)
            lon1_rad = math.radians(lon1)
            lat2_rad = math.radians(lat2)
            lon2_rad = math.radians(lon2)
            
            dlat = lat2_rad - lat1_rad
            dlon = lon2_rad - lon1_rad
            
            a = math.sin(dlat/2)**2 + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon/2)**2
            c = 2 * math.asin(math.sqrt(a))
            
            return R * c
        
        distance_km = haversine_distance(previous_lat, previous_lng, current_lat, current_lng)
        
        # Calculate required speed (km/h)
        time_hours = time_diff / 3600
        required_speed = distance_km / time_hours if time_hours > 0 else float('inf')
        
        # Maximum reasonable speed (accounting for flights)
        max_speed = 900  # km/h (commercial aircraft speed)
        
        if required_speed > max_speed:
            return False, f"Impossible movement: {distance_km:.1f}km in {time_hours:.1f}h requires {required_speed:.1f}km/h"
        
        return True, f"Movement pattern valid: {distance_km:.1f}km in {time_hours:.1f}h"
    
    def verify_activity(self, activity_data: Dict[str, Any]) -> Dict[str, Any]:
        """Main verification function that combines all checks"""
        
        lat = activity_data.get("gps_lat", 0)
        lng = activity_data.get("gps_long", 0)
        signal_strength = activity_data.get("signal_strength", -100)
        ip_address = activity_data.get("ip_address", "")
        previous_lat = activity_data.get("previous_lat")
        previous_lng = activity_data.get("previous_lng")
        time_diff = activity_data.get("time_diff", 3600)
        
        verification_result = {
            "verified": False,
            "score": 0,
            "max_score": 100,
            "checks": {},
            "timestamp": int(time.time()) if 'time' in globals() else 0
        }
        
        # GPS Coordinate Verification (30 points)
        gps_valid, gps_reason = self.verify_gps_coordinates(lat, lng)
        verification_result["checks"]["gps_validation"] = {
            "passed": gps_valid,
            "reason": gps_reason,
            "points": 30 if gps_valid else 0
        }
        if gps_valid:
            verification_result["score"] += 30
        
        # IP Reputation Check (40 points)
        if ip_address:
            ip_valid, ip_reason = self.check_ip_reputation(ip_address)
            verification_result["checks"]["ip_reputation"] = {
                "passed": ip_valid,
                "reason": ip_reason,
                "points": 40 if ip_valid else 0
            }
            if ip_valid:
                verification_result["score"] += 40
        else:
            verification_result["checks"]["ip_reputation"] = {
                "passed": False,
                "reason": "No IP address provided",
                "points": 0
            }
        
        # Signal Strength Verification (20 points)
        signal_valid, signal_reason = self.verify_signal_strength(signal_strength, lat, lng)
        verification_result["checks"]["signal_strength"] = {
            "passed": signal_valid,
            "reason": signal_reason,
            "points": 20 if signal_valid else 0
        }
        if signal_valid:
            verification_result["score"] += 20
        
        # Movement Pattern Analysis (10 points)
        movement_valid, movement_reason = self.analyze_movement_pattern(
            lat, lng, previous_lat, previous_lng, time_diff
        )
        verification_result["checks"]["movement_pattern"] = {
            "passed": movement_valid,
            "reason": movement_reason,
            "points": 10 if movement_valid else 0
        }
        if movement_valid:
            verification_result["score"] += 10
        
        # Overall verification (require 70% score)
        verification_result["verified"] = verification_result["score"] >= 70
        
        return verification_result

def main():
    """Main function for Switchboard execution"""
    try:
        # Parse input from Switchboard
        if len(sys.argv) > 1:
            # Input from command line argument
            input_data = json.loads(sys.argv[1])
        else:
            # Input from stdin
            input_data = json.loads(sys.stdin.read())
        
        # Initialize verifier
        verifier = ActivityVerifier()
        
        # Perform verification
        result = verifier.verify_activity(input_data)
        
        # Output result for Switchboard
        output = {
            "success": result["verified"],
            "data": result
        }
        
        print(json.dumps(output))
        
        # Return appropriate exit code
        sys.exit(0 if result["verified"] else 1)
        
    except Exception as e:
        error_output = {
            "success": False,
            "error": str(e),
            "data": None
        }
        print(json.dumps(error_output))
        sys.exit(1)

if __name__ == "__main__":
    main()
                        # Check for datacenter/hosting indicators
                        datacenter_keywords = ["hosting", "datacenter", "cloud", "server", "vps", "digital ocean", "aws", "azure", "linode"]
                        for keyword in datacenter_keywords:
                            if keyword in org:
                                return False, f"IP from datacenter/hosting provider: {org}"
                        
                        # Additional checks for proxy indicators
                        if data.get("proxy", False) or data.get("vpn", False):
                            return False, "IP flagged as proxy/VPN"
                            
                except Exception as e:
                    continue
            
            return True, "IP reputation check passed"
            
        except Exception as e:
            # If we can't verify, err on the side of caution but don't completely block
            return True, f"IP verification inconclusive: {str(e)}"
    
    def verify_signal_strength_plausibility(self, signal_strength: int, lat: float, lng: float) -> Tuple[bool, str]:
        """Verify signal strength is plausible for the location"""
        
        # Basic range check for WiFi signal strength
        if not (-100 <= signal_strength <= 0):
            return False, "Signal strength outside valid WiFi range (-100 to 0 dBm)"
        
        # Very strong signals (-10 to 0) are suspicious unless in urban areas
        if signal_strength > -10:
            # This would require more sophisticated location-based analysis
            # For now, we'll allow it but could add urban area detection
            pass
        
        # Very weak signals (-90 to -100) are also suspicious for earning rewards
        if signal_strength < -90:
            return False, "Signal strength too weak to be reliable"
        
        return True, "Signal strength within acceptable range"
    
    def calculate_distance_from_previous(self, lat1: float, lng1: float, lat2: float, lng2: float) -> float:
        """Calculate distance between two GPS points using Haversine formula"""
        
        R = 6371  # Earth's radius in kilometers
        
        lat1_rad = math.radians(lat1)
        lat2_rad = math.radians(lat2)
        delta_lat = math.radians(lat2 - lat1)
        delta_lng = math.radians(lng2 - lng1)
        
        a = (math.sin(delta_lat / 2) ** 2 + 
             math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(delta_lng / 2) ** 2)
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
        
        return R * c
    
    def verify_activity(self, activity_data: Dict[str, Any], user_ip: str, previous_location: Dict = None) -> Dict[str, Any]:
        """Main verification function"""
        
        result = {
            "verified": False,
            "reasons": [],
            "score": 0,
            "max_score": 100
        }
        
        try:
            lat = float(activity_data.get("gps_lat", 0))
            lng = float(activity_data.get("gps_long", 0))
            signal_strength = int(activity_data.get("signal_strength", -100))
            
            # GPS coordinate verification
            gps_valid, gps_reason = self.verify_gps_coordinates(lat, lng)
            if gps_valid:
                result["score"] += 30
            else:
                result["reasons"].append(f"GPS: {gps_reason}")
            
            # IP reputation check
            ip_valid, ip_reason = self.check_ip_reputation(user_ip)
            if ip_valid:
                result["score"] += 40
            else:
                result["reasons"].append(f"IP: {ip_reason}")
            
            # Signal strength verification
            signal_valid, signal_reason = self.verify_signal_strength_plausibility(signal_strength, lat, lng)
            if signal_valid:
                result["score"] += 20
            else:
                result["reasons"].append(f"Signal: {signal_reason}")
            
            # Movement pattern analysis (if previous location available)
            if previous_location:
                prev_lat = previous_location.get("lat", 0)
                prev_lng = previous_location.get("lng", 0)
                prev_timestamp = previous_location.get("timestamp", 0)
                current_timestamp = activity_data.get("timestamp", 0)
                
                distance = self.calculate_distance_from_previous(prev_lat, prev_lng, lat, lng)
                time_diff = current_timestamp - prev_timestamp
                
                # Check for impossible movement speeds (>500 km/h)
                if time_diff > 0:
                    speed_kmh = (distance / (time_diff / 3600))
                    if speed_kmh > 500:
                        result["reasons"].append(f"Movement: Impossible speed {speed_kmh:.1f} km/h")
                    else:
                        result["score"] += 10
            
            # Determine if verification passes (require 70% score)
            result["verified"] = result["score"] >= 70
            
            if not result["verified"]:
                result["reasons"].append(f"Overall score {result['score']}/{result['max_score']} below threshold")
            
        except Exception as e:
            result["reasons"].append(f"Verification error: {str(e)}")
        
        return result

def main():
    """Main function called by Switchboard"""
    
    try:
        # Read input data from Switchboard
        input_data = json.loads(sys.stdin.read())
        
        # Extract activity data and user IP
        activity_data = input_data.get("activity", {})
        user_ip = input_data.get("user_ip", "")
        previous_location = input_data.get("previous_location")
        
        # Initialize verifier and run verification
        verifier = ActivityVerifier()
        result = verifier.verify_activity(activity_data, user_ip, previous_location)
        
        # Prepare output for Switchboard
        output = {
            "verified": result["verified"],
            "score": result["score"],
            "reasons": result["reasons"],
            "timestamp": activity_data.get("timestamp"),
            "user_id": input_data.get("user_id")
        }
        
        # Output result as JSON
        print(json.dumps(output))
        
        # Return appropriate exit code
        sys.exit(0 if result["verified"] else 1)
        
    except Exception as e:
        error_output = {
            "verified": False,
            "error": str(e),
            "timestamp": None,
            "user_id": input_data.get("user_id") if 'input_data' in locals() else None
        }
        print(json.dumps(error_output))
        sys.exit(1)

if __name__ == "__main__":
    main()